src/librustc_typeck/check/mod.rs

```diff
@@ -2646,7 +2650,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     pub fn check_expr_has_type(&self,
                                expr: &'gcx hir::Expr,
                                expected: Ty<'tcx>) -> Ty<'tcx> {
-        let mut ty = self.check_expr_with_hint(expr, expected);
+        self.check_expr_expect_type(expr, ExpectHasType(expected))
+    }
+
+    fn check_expr_expect_type(&self,
```

Comment:
Maybe we can call this `check_expr_meets_expectation_or_error`? I think this zoo of functions is sort of confusing, and the existing naming convention isn't the clearest (for bonus points, we could apply the `_or_error` convention elsewhere, e.g. `check_expr_has_type_or_error`)